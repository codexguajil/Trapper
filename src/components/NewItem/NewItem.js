import React, { Component } from 'react';

export class NewItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      id: Date.now(),
      completed: false,
    }
  }

  handleItemChange = (e) => {
    this.setState({
      value: e.target.value
    })
    this.props.handleChange(e)
  }

  handleItemBlur = (e) => {
    let { id, value, completed } = this.state;
    let currItem = {id, value, completed}
    if(value) {
      this.props.handleItem(currItem)
    }
  }

  render() {
    return (
      <div className='list-container'>
        <button className='list-control'></button>
        <textarea value={this.state.value}
                  onChange={this.handleItemChange} 
                  onBlur={this.handleItemBlur}
                  name='listText'
                  placeholder='Add an item . . .'
                  className='list-item'
                >
        </textarea>
        <button className='list-control delete-item'>X</button>
      </div>
    )
  }
}

export default NewItem;