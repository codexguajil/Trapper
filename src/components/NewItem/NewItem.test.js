import NewItem from './NewItem'
import React from 'react'
import { shallow } from 'enzyme'

describe('NewItem', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      handleChange: jest.fn(),
      handleItem: jest.fn()
    }

    Date.now = jest.fn().mockImplementation(() => 6)
    wrapper = shallow(
      <NewItem {...props}/>
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual( {value: '', id: Date.now(), completed: false} )
  })

  it('should update value state onChange and call handleChange', () => {
    const mockState = {
      value: 'he',
      id: 6,
      completed: false,
    }

    const mockEvent = {
      target: {
        value: 'hey'
      }
    }

    wrapper.setState(mockState)
    expect(wrapper.instance().state).toEqual(mockState)

    wrapper.instance().handleItemChange(mockEvent)
    expect(wrapper.state('value')).toEqual('hey')
    expect(wrapper.instance().props.handleChange).toHaveBeenCalledWith(mockEvent)
  })

describe('handleItemBlur', () => {
  it('should call handleItem with the properties in state', () => {
    const mockState = {
      value: 'he',
      id: 6,
      completed: false,
    }
    wrapper.setState(mockState)
    expect(wrapper.instance().state).toEqual(mockState)

    wrapper.instance().handleItemBlur()
    expect(wrapper.instance().props.handleItem).toHaveBeenCalledWith(mockState)
  })

  it('should not call handleItem if the value in state is and empty string', () => {
    const mockState = {
      value: '',
      id: 6,
      completed: false,
    }
    wrapper.setState(mockState)
    expect(wrapper.instance().state).toEqual(mockState)

    wrapper.instance().handleItemBlur()
    expect(wrapper.instance().props.handleItem).not.toHaveBeenCalled()
  })
})
})