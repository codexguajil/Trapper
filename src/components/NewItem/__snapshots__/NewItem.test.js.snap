// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewItem should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewItem
    handleChange={[MockFunction]}
    handleItem={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="list-control"
        />,
        <textarea
          className="list-item"
          name="listText"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Add an item . . ."
          value=""
        />,
        <button
          className="list-control delete-item"
        >
          X
        </button>,
      ],
      "className": "list-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "list-control",
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "list-item",
          "name": "listText",
          "onBlur": [Function],
          "onChange": [Function],
          "placeholder": "Add an item . . .",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "X",
          "className": "list-control delete-item",
        },
        "ref": null,
        "rendered": "X",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="list-control"
          />,
          <textarea
            className="list-item"
            name="listText"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Add an item . . ."
            value=""
          />,
          <button
            className="list-control delete-item"
          >
            X
          </button>,
        ],
        "className": "list-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "list-control",
          },
          "ref": null,
          "rendered": null,
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "list-item",
            "name": "listText",
            "onBlur": [Function],
            "onChange": [Function],
            "placeholder": "Add an item . . .",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "X",
            "className": "list-control delete-item",
          },
          "ref": null,
          "rendered": "X",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
